frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -2748,6 +2762,10 @@ TypeInference::matchCase(const IR::SelectExpression* select, const IR::Type_Tupl
     // The caseType may be a simple type, and then we have to unwrap the selectType
     if (caseType->is<IR::Type_Dontcare>())
         return selectCase;
+    // The caseType may be an value set, we use the type of the selectType
```

Comment:
This looks wrong. You should handle it exactly like a tuple is handled below.
Just extract the type argument of the value_set and use that to match the select case type.